python/cuml/cuml/neighbors/kernel_density.py

```diff
@@ -167,19 +171,86 @@ class KernelDensity(Base):
+        estimated with ``bandwidth="scott"`` or ``bandwidth="silverman"``.
+
     Examples
     --------
+    >>> from cuml.neighbors import KernelDensity
+    >>> import cupy as cp
+    >>> rng = cp.random.RandomState(42)
+    >>> X = rng.random_sample((100, 3))
+    >>> kde = KernelDensity(kernel='gaussian', bandwidth=0.5).fit(X)
+    >>> log_density = kde.score_samples(X[:3])
+    """
 
-    .. code-block:: python
+    _cpu_class_path = "sklearn.neighbors.KernelDensity"
 
-        >>> from cuml.neighbors import KernelDensity
-        >>> import cupy as cp
-        >>> rng = cp.random.RandomState(42)
-        >>> X = rng.random_sample((100, 3))
-        >>> kde = KernelDensity(kernel='gaussian', bandwidth=0.5).fit(X)
-        >>> log_density = kde.score_samples(X[:3])
+    @classmethod
+    def _get_param_names(cls):
+        return [
+            *super()._get_param_names(),
+            "bandwidth",
+            "kernel",
+            "metric",
+            "metric_params",
+        ]
 
-    """
+    @classmethod
+    def _params_from_cpu(cls, model):
+        if model.metric not in SUPPORTED_METRICS:
+            raise UnsupportedOnGPU(
+                f"`metric={model.metric!r}` is not supported"
+            )
+        return {
+            "bandwidth": model.bandwidth,
+            "kernel": model.kernel,
+            "metric": model.metric,
+            "metric_params": model.metric_params,
+        }
+
+    def _params_to_cpu(self):
+        return {
+            "bandwidth": self.bandwidth,
+            "kernel": self.kernel,
+            "metric": self.metric,
+            "metric_params": self.metric_params,
+        }
+
+    def _attrs_from_cpu(self, model):
+        X = cp.asarray(model.tree_.data, dtype=cp.float64)
+        sample_weight = (
+            None
+            if model.tree_.sample_weight is None
+            else cp.asarray(model.tree_.sample_weight, dtype=cp.float32)
+        )
+        return {
+            "bandwidth_": model.bandwidth_,
+            "_X": X,
+            "_sample_weight": sample_weight,
+            **super()._attrs_from_cpu(model),
+        }
+
+    def _sync_attrs_to_cpu(self, model):
+        # There's no way to create a `sklearn.neighbors.KernelDensity` instance
+        # from our stored state without fitting a new KernelDensity (since
+        # sklearn always uses a pre-fit tree rather than a brute-force method).
+        # As such, syncing to CPU is effectively just a refit. Fitting isn't
+        # that expensive here, and unlike other estimators it's not clear that
+        # a fit-on-gpu, predict-on-cpu model is useful given typical
+        # KernelDensity use cases. What we have here should be fine for now.
+        X = cp.asnumpy(self._X)
+        sample_weight = (
+            None
+            if self._sample_weight is None
+            else cp.asnumpy(self._sample_weight)
+        )
+        model.fit(X, sample_weight=sample_weight)
```

Comment:
I agree that this is seemingly the only option to perform the conversion due to the architectural differences. I'd consider this reasonable given the expected use characteristics.